### PR TITLE
Tbox 73 cloud transfer

### PR DESCRIPTION
**Pull request by DIvkov575**
# ↪️ Pull Request

## ✔️ PR Todo

- [x] Add documentation
- [x] Add examples
- [x] Ensure you are following the practices in the [contributor guide](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit) and [coding_standards](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards)
  * [Use consistent and descriptive name conventions](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.477s36w4rds9) 
  * [Uses google-style docstrings](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Google-style-docstrings)
  * [Uses versioned API/client methods only](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.yv6df99fyrq2)
  * [Avoid global variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-global-variables)
  * [Avoid giant try/catch](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-giant-try-/-except)
  * [Don't use mutable default variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Don%E2%80%99t-use-mutable-default-arguments)
  * [Use a `main()` function in your scripts](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Use-a-%E2%80%9Cmain()%E2%80%9D--function-in-your-scripts)
- [x] Added/updated testing for this change (ensure you are testing any changes/new code!)
- [x] Update title to link to TBOX issue (i.e. start pull request title with TBOX-xx) 
- [ ] Approval from first code-reviewer
- [ ] Approval from second code-reviewer (only required if a second reviewer has left comments)
- [x] Pass all checks

### CODE REVIEW

**Comment by ycharm** [comment]
Hi we had a docs bug that was blocking the tests from passing, can you merge in develop to your PR. It will allow the tests to pass. Thanks!

**Comment by DIvkov575** [comment]
@jun-dao What would you recommend for testing google & AWS uploading/downloading? Thank you.

File: requirements.txt
Comment (Kovacsova) [inline requirements.txt]: The cloud requirements are only needed in optional_requirements.txt

File: tamr_toolbox/data_io/__init__.py
Comment (Kovacsova) [inline tamr_toolbox/data_io/__init__.py:4]: I suppose csv was removed accidentally?

Comment (DIvkov575) [reply tamr_toolbox/data_io/__init__.py:4]: yes

**Comment by ycharm** [comment]
> 

hi any progress on trying to do testing? Can I help in any way?

**Comment by DIvkov575** [comment]
For now, I'm just looking into the boto3 and botocore libraries. Im not sure whether to try and use `moto` for mocking because it is not a tamr dependency.

**Comment by DIvkov575** [comment]
@ycharm  Im getting an issue with asserting that a temp file is a tarfile. Any recommendations for getting around this?
Is there anything else I should add to the test?

**Comment by ycharm** [comment]
> @ycharm Im getting an issue with asserting that a temp file is a tarfile. Any recommendations for getting around this? Is there anything else I should add to the test?
Is this helpful: https://stackoverflow.com/questions/23883947/combining-tempfile-and-tarfile-with-python

**Comment by DIvkov575** [comment]
@ycharm I'm realizing that the test would not work through trying to unit test mock it. The mock does not return anything so asserting is impossible on an empty temp file.

I've still been unable to find any alternative default credentials or mocking for GCP, but have an idea for testing. 

testing tarring could be done by adding an optional return to the cloud_upload function (returns file about to be uploaded).
Is this a viable option for testing?

**Comment by ycharm** [comment]
@DIvkov575 I don't see anything in our guidelines that suggests we shouldn't have the optional return. Especially if you need it for testing. @clint-richardson check out the question please.

**Comment by ycharm** [comment]
Hi It looks like the tarring is causing tests to fail. I'll see if I can find a non-local/mock-api test for it and also will try to test your functions out.

**Comment by DIvkov575** [comment]
@ycharm  sorry bout the slow reply. I'm trying out a mock I made for the untarring of the test right now. It seems to be working for the most part.

File: requirements.txt
Comment (skalish) [inline requirements.txt]:
```suggestion
```
Please remove this line. This package is only needed as part of the `optional_requirements.txt` file.

**Comment by skalish** [comment]
Since this functionality is about moving file between a local filesystem and cloud storage, and not for data movement into and out of Tamr, it would be more appropriate for the `cloud` module to be under the top-level `filesystem` module, rather than `data_io` (which is specifically for ingesting/exporting Tamr data).

File: examples/snippets/data_io/cloud/gcs_upload.py
Comment (skalish) [inline examples/snippets/data_io/cloud/gcs_upload.py]: These example snippets should pass appropriate arguments to the functions.  Here `cloud_client` should be whatever GCS Python client object would need to be used.  It isn't necessary that the object be fully set up, but it should demonstrate what other package the user needs.

File: optional_requirements.txt
Comment (skalish) [inline optional_requirements.txt]: I think you have the wrong package imported here.  It looks like `google` is related to the search engine, while `google-cloud-storage` is the installation name for the client that includes cloud storage.  Confusingly, you `import google` with either of them.
```suggestion
google-cloud-storage>=2.4.0
```

File: examples/snippets/data_io/cloud/gcs_upload.py
Comment (skalish) [reply examples/snippets/data_io/cloud/gcs_upload.py]: I think the object you want here is a `google.cloud.storage.Client`.  I would recommend you include `from google.cloud.storage import Client as GcsClient` at the top, and then create ` gcs_client = Client()` to pass into the function.
Then it would be best to show how the `destination_filepath` does not include the bucket name, by including a comment along the lines of `# upload a local file "my_local_directory/my_file.txt" to "gs://my-bucket/path-to-file"` and then passing the appropriate strings to `destination_filepath` and `bucket_name`.

File: tamr_toolbox/data_io/cloud.py
Comment (skalish) [inline tamr_toolbox/data_io/cloud.py]: This and the rest of the arguments should be correctly type-hinted.
You can `from google.cloud.storage import Client as GcsClient` at the top of the file, then
```suggestion
    cloud_client: "GcsClient",
```
The type needs to be quoted since this is an optional dependency.

File: tamr_toolbox/data_io/cloud.py
Comment (skalish) [inline tamr_toolbox/data_io/cloud.py]: If I understand correctly, the `tar_file` functionality is just for tarring up a file before uploading it (and for download functions is just about untarring a downloaded file).

This is cool functionality that is probably useful, since it allows uploading a collection of files, but in this implementation it over-complicated the code.

Can you simplify the upload and download functions so that they *only* upload/download a single file?  The tar/untar should be its own function, which could be used in combination with the simpler upload/download to have the same effect with more readability.

File: tamr_toolbox/data_io/cloud.py
Comment (skalish) [inline tamr_toolbox/data_io/cloud.py]: Similar to GCS, you'll need to `from mypy_boto3_s3.client import S3Client` and type hint like this:
```suggestion
    cloud_client: S3Client,
```

File: optional_requirements.txt
Comment (skalish) [inline optional_requirements.txt:9]: In addition to `boto3` we will need to add `boto3-stubs-lite[essential]>=1.21.21` as an optional dependency to support type-hinting.

File: tamr_toolbox/data_io/cloud.py
Comment (skalish) [inline tamr_toolbox/data_io/cloud.py]: To appropriately type hint with optional dependencies you will need to include the following at the top of this file (after imports): 
```python
# Building our documentation requires access to all dependencies, including optional ones
# This environments variable is set automatically when `invoke docs` is used
BUILDING_DOCS = os.environ.get("TAMR_TOOLBOX_DOCS") == "1"
if BUILDING_DOCS:
    # Import relevant optional dependencies
    from google.cloud.storage import Client as GcsClient
    from mypy_boto3_s3.client import S3Client
```

File: setup.py
Comment (skalish) [inline setup.py]: Related to the addition to `optional_dependencies.txt`, `boto3-stubs-lite[essential]>=1.21.21` will need to be added here as well.

File: tamr_toolbox/data_io/cloud.py
Comment (DIvkov575) [reply tamr_toolbox/data_io/cloud.py]: What type's should I add for the generic upload/download functions?

Comment (DIvkov575) [reply tamr_toolbox/data_io/cloud.py]: I was also wondering if adding the condition is necessary because with the condition in place, the condition causes NameError when testing for dependencies.

File: tamr_toolbox/filesystem/cloud.py
Comment (skalish) [inline tamr_toolbox/filesystem/cloud.py]: I don't recall why, but this doesn't work for this module.  You'll need to `from google.cloud import storage` and then refer to `storage.Client`.

Comment (skalish) [reply tamr_toolbox/filesystem/cloud.py]: This is the pattern in the documentation for that package, so it seems reasonable to follow it: https://pypi.org/project/google-cloud-storage/

File: tamr_toolbox/filesystem/cloud.py
Comment (skalish) [inline tamr_toolbox/filesystem/cloud.py]: I don't think this cloud-agnostic function or the `file_download` function are necessary.  It feels like a lot of code to just pass a string and use a different function.  Users can just directly use e.g. `s3_download`.

File: doc_src/examples/data_io.rst
Comment (skalish) [inline doc_src/examples/data_io.rst]: It looks like the CSV block here got dropped accidentally.  Let me know if it would help to go over any git issues together.

File: tamr_toolbox/filesystem/cloud.py
Comment (skalish) [inline tamr_toolbox/filesystem/cloud.py]: Because these are optional dependencies, they should be imported in a slightly different way:
```suggestion
# Building our documentation requires access to all dependencies, including optional ones
# This environments variable is set automatically when `invoke docs` is used
BUILDING_DOCS = os.environ.get("TAMR_TOOLBOX_DOCS") == "1"
if BUILDING_DOCS:
    # Import relevant optional dependencies
    from google.cloud import storage
    from mypy_boto3_s3.client import S3Client
```

This also means references to these modules in type-hints need to be in double quotes. See my other comment.

File: tamr_toolbox/filesystem/cloud.py
Comment (skalish) [inline tamr_toolbox/filesystem/cloud.py]: With the previous suggested change, type hints involving the optional modules need to be specified as strings like so:
```suggestion
    cloud_client: "storage.Client",
```

Comment (skalish) [reply tamr_toolbox/filesystem/cloud.py]: This will also need to be done in the other function definitions in this file, including the S3 ones.

File: setup.py
Comment (ycharm) [inline setup.py]: is this a typo?

Comment (DIvkov575) [reply setup.py]: this looks the same as the develop branch

File: examples/snippets/filesystem/cloud/gcs_upload.py
Comment (wpmartin) [inline examples/snippets/filesystem/cloud/gcs_upload.py]: Can this filepath be made consistent with the other examples? This is the only example to use the bucket and "gs://" prefix for the gcs filepath

File: setup.py
Comment (wpmartin) [inline setup.py]: typo in "require"

File: setup.py
Comment (skalish) [reply setup.py]: I think the typo referred to here is in a following line.  There is another comment pointing it out.

File: examples/snippets/filesystem/cloud/gcs_upload.py
Comment (DIvkov575) [reply examples/snippets/filesystem/cloud/gcs_upload.py]: The s3 upload functions take a path within the bucket. Should I change the function to remove the prefix from a 's3 link'?

File: tamr_toolbox/filesystem/cloud.py
Comment (skalish) [inline tamr_toolbox/filesystem/cloud.py]: These parameters still need type hinting and should not have default values.  It would also be best to force their use as keyword-only by using the `*` parameter separator:
```suggestion
    *,
    source_filepath: str,
    destination_filepath: str,
    bucket_name: str,
```

File: tamr_toolbox/filesystem/cloud.py
Comment (skalish) [inline tamr_toolbox/filesystem/cloud.py]:
```suggestion
    cloud_client: "storage.Client",
    *,
    source_filepath: str,
    destination_filepath: str,
    bucket_name: str,
```

File: tamr_toolbox/filesystem/cloud.py
Comment (skalish) [inline tamr_toolbox/filesystem/cloud.py]:
```suggestion
    cloud_client: "S3Client",
    *,
    source_filepath: str,
    destination_filepath: str,
    bucket_name: str,
```

File: tamr_toolbox/filesystem/cloud.py
Comment (skalish) [inline tamr_toolbox/filesystem/cloud.py]:
```suggestion
    cloud_client: "S3Client",
    *,
    source_filepath: str,
    destination_filepath: str,
    bucket_name: str,
```

File: examples/snippets/filesystem/cloud/gcs_download.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/gcs_download.py]: My testing shows that the `source_filepath` should not include `gs://<bucket-name>` and should just be the component of the filepath following this.
```suggestion
    source_filepath="path-to-file",
```

File: examples/snippets/filesystem/cloud/gcs_download.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/gcs_download.py]: This is not the correct module to import
```suggestion
from google.cloud import storage

gcs_client = storage.Client()
```

File: examples/snippets/filesystem/cloud/gcs_upload.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/gcs_upload.py]: This is not the correct module to import
```suggestion
from google.cloud import storage

gcs_client = storage.Client()
```

File: examples/snippets/filesystem/cloud/gcs_upload.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/gcs_upload.py]:
```suggestion
    destination_filepath="path-to-file",
```

File: examples/snippets/filesystem/cloud/gcs_download.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/gcs_download.py]:
```suggestion
An example script to demonstrate the use of the gcs_download function to download file from a GCS bucket.
```

File: examples/snippets/filesystem/cloud/s3_download.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/s3_download.py]:
```suggestion
An example script to demonstrate the use of the s3_download function to download file from an S3 bucket.
```

File: examples/snippets/filesystem/cloud/s3_upload.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/s3_upload.py]: This works, but is not what is recommended in the boto3 docs [here](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-uploading-files.html).  Instead I see:
```suggestion
import boto3

s3_client = boto3.client("s3")
```

File: examples/snippets/filesystem/cloud/s3_upload.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/s3_upload.py]:
```suggestion
    cloud_client=s3_client,
```

File: examples/snippets/filesystem/cloud/s3_download.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/s3_download.py]:
```suggestion
import boto3

s3_client = boto3.client("s3")

# download file from AWS S3
# download a file on GCS "s3://my-bucket/path-to-file" to "my_local_directory/my_file.txt"
tbox.filesystem.cloud.s3_download(
    cloud_client=s3_client,
```

File: examples/snippets/filesystem/cloud/s3_download.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/s3_download.py]: My testing shows that the `source_filepath` should not include `s3://<bucket-name>` and should just be the component of the filepath following this.
```suggestion
    source_filepath="path-to-file",
```

File: examples/snippets/filesystem/cloud/s3_upload.py
Comment (skalish) [inline examples/snippets/filesystem/cloud/s3_upload.py]:
```suggestion
    destination_filepath="path-to-file",
```

**Review by skalish — CHANGES_REQUESTED** [review]

**Comment by DIvkov575** [comment]
@skalish are there any other changes that I should make?

File: setup.py
Comment (skalish) [inline setup.py:44]: We should also set individual sets for each of the clouds.  I think `s3` and `gcs` are good names:
```suggestion
        "ssh": ["paramiko>=2.8.0"],
        "gcs": ["google-cloud-storage>=2.0.0"],
        "s3": ["boto3>=1.21.21", "boto3-stubs-lite[essential]>=1.21.21"],
```

File: setup.py
Comment (skalish) [inline setup.py]: Is it necessary to explicitly install the `google` package?

File: tests/filesystem/test_cloud.py
Comment (skalish) [inline tests/filesystem/test_cloud.py]:
```suggestion
```

File: tests/filesystem/test_cloud.py
Comment (skalish) [inline tests/filesystem/test_cloud.py]:
```suggestion
```

File: tests/filesystem/test_cloud.py
Comment (skalish) [inline tests/filesystem/test_cloud.py]:
```suggestion
```

File: tests/filesystem/test_cloud.py
Comment (skalish) [inline tests/filesystem/test_cloud.py]:
```suggestion
```
Here you are testing something about the underlying package that your function calls.  This isn't necessary, since we only need to test our own functionality.  This part of the test also makes it hard for other developers to test locally, since their environments might have valid default credentials that will get picked up by the `google.cloud.client.Client()` call.

File: setup.py
Comment (skalish) [inline setup.py]:
```suggestion
```
I don't think it is necessary to explicitly install the `google` package.  Have you seen that this isn't the case?

File: tests/filesystem/test_cloud.py
Comment (skalish) [inline tests/filesystem/test_cloud.py]: These tests are not actually checking that the function does anything, only that it doesn't raise errors.  You can take them further by asserting that the underlying package's functions are being called like this:
```suggestion
            )
            
            bucket = client.get_bucket
            bucket.assert_called_with("test-bucket")

            blob = bucket().blob
            blob.assert_called_with("path_in_bucket.txt")

            download = blob().download_to_filename
            download.assert_called_with(download_tmp.name)
```

Please also make similar changes to the other 3 tests.

File: tests/filesystem/test_cloud.py
Comment (skalish) [inline tests/filesystem/test_cloud.py]: `patch` can be used in a more simple way by just using it on your test function. Then you don't need to have this extra function `test_download`:
```suggestion
@patch("google.cloud.client.Client")
def test_gcs_download(mock_client):
    client = mock_client()

    with tempfile.NamedTemporaryFile() as download_tmp:
        cloud.gcs_download(
            cloud_client=client,
            source_filepath="path_in_bucket.txt",
            destination_filepath=download_tmp.name,
            bucket_name="test-bucket",
        )
```

**Review by skalish — CHANGES_REQUESTED** [review]
There are a couple of dependency definitions to be taken care of.  The testing should also be reworked to be a bit cleaner and more complete.  I think those will be the last needed changes.  Let me know if I can help at all.

File: setup.py
Comment (DIvkov575) [reply setup.py]: As strange as it is, I think importing both is necessary. (google-cloud-storage & google)

Comment (DIvkov575) [reply setup.py]: The docs import the google cloud storage from google.
https://pypi.org/project/google-cloud-storage/

Comment (skalish) [reply setup.py]: Does not including it cause `invoke docs` to fail for you? I'm not running into that issue.  For me it looks like installing `google-cloud-translate==3.7.4` and `google-cloud-storage==2.0.0` is enough to run all tests and build docs.

Comment (DIvkov575) [reply setup.py]: ok, ill change that. I do not remember what the issue was, we'll see if it's raised again.

Comment (skalish) [reply setup.py]: Did you try removing this and see whether there was a problem?

File: tamr_toolbox/data_io/__init__.py
Comment (skalish) [inline tamr_toolbox/data_io/__init__.py:4]: It looks like this accidental line removal happened again. Can you restore it?

**Review by skalish — CHANGES_REQUESTED** [review]
After you resolve the one accidental import removal, this is good to go.  I also asked a question about the need for the `google` package.  If you can follow-up on that it would be nice to confirm whether it is necessary, but that isn't absolutely needed.  I'll approve as soon as I can once these last changes are made.

**Comment by DIvkov575** [comment]
> After you resolve the one accidental import removal, this is good to go. I also asked a question about the need for the `google` package. If you can follow-up on that it would be nice to confirm whether it is necessary, but that isn't absolutely needed. I'll approve as soon as I can once these last changes are made.

I think that the google import / install is necessary because inorder to use the google-cloud-storage client you need to import it with `from google.cloud import storage` and without google installed it wont be able to find where to import from

**Review by skalish — APPROVED** [review]

**Comment by skalish** [comment]
We shouldn't spend more time digging in on the `google` package question. It will be very easy to remove if needed in the future.  This all looks good to me!  Thanks for all of the work you've put it on this.

Let me know if you'd like me to click to merge this in for you when you're ready.

**Comment by DIvkov575** [comment]
@skalish yes, could you merge this. I don't have a merge button.

**Pull request merged**
